### ISSUE TITLE
Porting chromium's clang plugin to flutter

### ISSUE THREAD

**Issue by 0xZOne** [flutter/flutter#93039]
## Use case

Related issue: https://github.com/flutter/engine/pull/29402#issuecomment-958979552

## Proposal

Chromium has a clang plugin that detects ordering of WeakPtrFactory ivars. Maybe we should port it to flutter.

**Comment by dnfield** [comment]
Ah, posted this on the issue instead of here - Talked to some Fuchsia Toolchain folks, and they suggested implementing these as clang-tidy checks. Fuchsia already has some at https://github.com/llvm/llvm-project/tree/main/clang-tools-extra/clang-tidy/fuchsia

I am not clear on who might be available to do this, if anyone is right now, or if this is something @0xZOne would be interestedin trying, but just trying to leave some more breadcrumbs here.

/cc @petrhosek

**Comment by akbiggs** [comment]
@dnfield If the suggestion is to make this check part of a ClangTidyModule in the clang-tidy codebase (as Fuchsia does in that folder) I'm worried that the llvm/clang-tidy team will reject that PR. WeakPtrFactory is only defined as an internal detail of the Flutter Engine repo so it would only be useful to people running clang_tidy in a clone of our single repo. The ClangTidyModules that exist in clang-tidy all seem targeted for public libraries/SDKs that can be used in many repos such as boost, the fuchsia public SDK, etc.

Writing the check as a clang plugin as @0xZOne is suggesting seems better-suited to me. The Chromium documentation mentions that [it could be unstable](https://chromium.googlesource.com/chromium/src/+/main/docs/writing_clang_plugins.md#Don_t-write-a-clang-plugin) but it seems to match their valid use cases for a plugin (we're writing a check that's only useful in the Flutter Engine codebase).

**Comment by dnfield** [comment]
Would clang-tidy support simply specifying a data member that must come last in a class definition?

The part I'm not clear on is if we do this as a plugin, do we have to do it as part of our clang build or not.

**Comment by 0xZOne** [comment]
> (we're writing a check that's only useful in the Flutter Engine codebase).

Is there a release plan? Very much looking forward to it.  @akbiggs

**Comment by akbiggs** [comment]
@0xZOne Sorry, I misspoke. I meant that IF we write a clang plugin then we are writing a plugin that is only useful in the Flutter Engine codebase.

@dnfield and I spoke with @petrhosek and it sounded like we could write a ClangTidy check that uses the [.clang-tidy file](https://cs.opensource.google/fuchsia/fuchsia/+/main:.clang-tidy;l=43) to configure what members need to come last in a class in a given codebase. This would be generally useful.

I'm not sure how this check should behave when there is more than one field that needs to go last in the same file though.

**Comment by dnfield** [comment]
> I'm not sure how this check should behave when there is more than one field that needs to go last in the same file though.

It should just fail right?

**Comment by 0xZOne** [comment]
> > I'm not sure how this check should behave when there is more than one field that needs to go last in the same file though.
> 
> It should just fail right?

Chromium has considered this issue, as below for detail:

[Clang plugin: Allow >1 WeakPtrFactory member in a class (Closed)](https://codereview.chromium.org/740083002)

However, If the two fields are dependent, their order still may need to be guaranteed by the developer.

**Comment by 0xZOne** [comment]
> > I'm not sure how this check should behave when there is more than one field that needs to go last in the same file though.
> 
> It should just fail right?

Chromium has considered this issue, as below for detail:

[Clang plugin: Allow >1 WeakPtrFactory member in a class (Closed)](https://codereview.chromium.org/740083002)

However, If the two fields are dependent, their order still may need to be guaranteed by the developer.

**Comment by akbiggs** [comment]
Should we be blocking the submission of https://github.com/flutter/engine/pull/29402 on this? It seems like that PR can help potentially reduce flakes if destroying WeakPtrFactory after any other members is dangerous, and we don't have an assignee for this currently.

**Comment by 0xZOne** [comment]
Yes, I totally agree with @akbiggs. Maybe we should consider merging this PR first.

**Comment by dnfield** [comment]
I've requested a test exemption on that PR.

**Comment by dnfield** [comment]
Fuschia toolchain team tells us they can work on this but don't have a timeline just yet. It will take a little while. In the mean time we should fix up the ordering while we wait for this to come in.

**Comment by petrhosek** [comment]
I have filed an issue [fxbug.dev/88447](https://bugs.fuchsia.dev/p/fuchsia/issues/detail?id=88447) on the Fuchsia side to track this.
